exp/services/ledgerexporter/internal/resumablemanager.go

```
@@ -0,0 +1,121 @@
+			// requested to start at a point beyond the latest network, resume not applicable.
+			return 0, false
+		}
+	}
+
+	binarySearchStop := max(end, networkLatest)
+	binarySearchStart := start
+
+	log.Infof("Resumability is searching datastore for next absent object key of requested export ledger range")
+
+	rangeSize := max(int(binarySearchStop-binarySearchStart), 1)
+	lowestAbsentIndex := sort.Search(rangeSize, binarySearchCallbackFn(&rm, ctx, binarySearchStart, binarySearchStop))
+	if lowestAbsentIndex < 1 {
+		// data store had no data within search range
+		return 0, false
+	}
+
+	if lowestAbsentIndex < int(rangeSize) {
+		nearestAbsentLedgerSequence := binarySearchStart + uint32(lowestAbsentIndex)
+		log.Infof("Resumability determined next absent object start key of %d for requested export ledger range", nearestAbsentLedgerSequence)
+		return nearestAbsentLedgerSequence, false
+	}
+
+	// unbounded, and datastore had up to latest network, return that as staring point.
+	if networkLatest > 0 {
+		return networkLatest, false
+	}
+
+	// data store had all ledgers for requested range, no resumability needed.
+	log.Infof("Resumability found no absent object keys in requested ledger range")
+	return 0, true
+}
+
+func binarySearchCallbackFn(rm *resumableManagerService, ctx context.Context, start, end uint32) func(ledgerSequence int) bool {
+	lookupCache := map[string]bool{}
+
+	return func(binarySearchIndex int) bool {
+		objectKeyMiddle := rm.config.LedgerBatchConfig.GetObjectKeyFromSequenceNumber(start + uint32(binarySearchIndex))
+
+		// there may be small occurrence of repeated queries on same object key once
+		// search narrows down to a range that fits within the ledgers per file
+		// worst case being 'log of ledgers_per_file' queries.
+		middleFoundOnStore, foundInCache := lookupCache[objectKeyMiddle]
+		if !foundInCache {
+			var datastoreErr error
+			middleFoundOnStore, datastoreErr = rm.dataStore.Exists(ctx, objectKeyMiddle)
+			if datastoreErr != nil {
```

If we encounter an error we should have a retry policy. If after retrying the request still fails we should propagate that error back to the caller of `FindStart()`.

---

ok, this touches on same regards to propagating errors from resumability rather than swallowing as part of a [best effort, seamless intent](https://github.com/stellar/go/pull/5264/files/7ddc4a0dc7ea8e9248b6f46e003db3a90b6af877#r1581254966), I don't want to add retry logic here for now, more complexity, instead I will add error to the return value here now, as there are a few new comments here which all converge on this, thanks!

---

sure, that sounds good. please create a new ticket for adding retry logic because, given that a full history data lake will require about 25 GCS requests to find the latest ledger, it can be likely that we encounter a transient error on occasion.

---

https://stellarorg.atlassian.net/browse/HUBBLE-392